pkg/monitor/main.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"strconv"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/daemon/defaults"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/byteorder"
+	"github.com/cilium/cilium/pkg/monitor/payload"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const targetName = "cilium-node-monitor"
+
+func receiveEvent(m *bpf.PerfEventSample, c int) {
+	pl := payload.Payload{Data: m.DataCopy(), CPU: c, Lost: 0, Type: payload.EventSample}
+	send(pl)
+}
+
+func lostEvent(m *bpf.PerfEventLost, c int) {
+	pl := payload.Payload{Data: []byte{}, CPU: c, Lost: m.Lost, Type: payload.RecordLost}
+	send(pl)
+}
+
+var (
+	listeners     = []net.Conn{}
+	mutex         sync.Mutex
+	monitorEvents *bpf.PerCpuEvents
+)
+
+// handleMonitorConn handles all the incoming monitors.
+func handleMonitorConn(server net.Listener) {
+	for {
+		conn, err := server.Accept()
+		if err != nil {
+			log.Print(err)
+			continue
+		}
+
+		mutex.Lock()
+		listeners = append(listeners, conn)
+		log.Infof("New monitor connected. len(listeners) = %d", len(listeners))
+		mutex.Unlock()
+	}
+}
+
+func dumpStat() {
+	c := int64(monitorEvents.Cpus)
+	n := int64(monitorEvents.Npages)
+	p := int64(monitorEvents.Pagesize)
+	l, u := monitorEvents.Stats()
+	m := models.MonitorStatus{Cpus: c, Npages: n, Pagesize: p, Lost: int64(l), Unknown: int64(u)}
+
+	mp, err := json.Marshal(m)
+	if err != nil {
+		log.Errorf("json: %s", err)
+		return
+	}
+	fmt.Println(string(mp))
+}
+
+// runMonitor starts monitoring.
+func runMonitor(npages int) {
+	log.Info("Starting monitoring traffic")
+	c := bpf.DefaultPerfEventConfig()
+	c.NumPages = npages
+
+	me, err := bpf.NewPerCpuEvents(c)
+	if err != nil {
+		log.Errorf("Error while starting monitor: %s", err)
+		return
+	}
+	monitorEvents = me
+
+	// Dump stat
+	t := time.NewTicker(5 * time.Second)
+	go func() {
+		for {
```

Comment:
The `for { select{ case:` that you wrote is doing the same thing that `for _ := range <- t.C` is doing